File: src/ir/symbolic/property.ts
/**
 * Property and bridge collectors — symbolic primitives that capture IR
 * declarations into a thread-local collection.
 *
 * The user's invariant file imports `property` and `bridge` from this
 * module. When the file runs (e.g., `await import("./parseInt.invariant.ts")`),
 * each call to `property()` and `bridge()` adds an entry to the active
 * collector. The lifter then reads the collector's contents.
 *
 * No tsc compiler API. No AST walking. The user's code RUNS to produce IR.
 *
 * This is the meta-circular property of the framework: the IR comes from
 * RUNNING the invariant code, not from COMPILING it. Each kit's
 * symbolic primitives provide the IR-emission mechanism for that host
 * language.
 */

import type { IrFormula, IrTerm, Sort } from "../formulas.js";

// ---------------------------------------------------------------------------
// Declarations captured by the collector
// ---------------------------------------------------------------------------

export interface PropertyDeclaration {
  kind: "property";
  name: string;
  formula: IrFormula;
}

export interface BridgeDeclaration {
  kind: "bridge";
  name: string;
  sourceSymbol: string;
  sourceLayer: string;
  targetContractCid: string;
  targetLayer: string;
  notes?: string;
}

export type Declaration = PropertyDeclaration | BridgeDeclaration;

// ---------------------------------------------------------------------------
// Active collector — module-scoped, set by the lifter before importing
// the user's invariant file.
// ---------------------------------------------------------------------------

let activeCollector: Declaration[] | null = null;

/**
 * Begin collecting declarations. Returns a function that finalizes
 * collection and returns the captured declarations.
 *
 * Usage (lifter):
 *   const finish = beginCollecting();
 *   await import("./user-invariants.invariant.ts");
 *   const declarations = finish();
 */
export function beginCollecting(): () => Declaration[] {
  if (activeCollector !== null) {
    throw new Error(
      "beginCollecting: another collection is already active; lifting is not re-entrant",
    );
  }
  const collector: Declaration[] = [];
  activeCollector = collector;
  return () => {
    activeCollector = null;
    return collector;
  };
}

/** Test helper: clear any active collector (use only in test setup/teardown). */
export function _resetCollector(): void {
  activeCollector = null;
  describePath = [];
  // Reset the quantifier variable counter so successive runs of the
  // same invariant code produce identical IR (and therefore identical
  // CIDs). The counter generates fresh variable names like `_x0`, `_x1`;
  // without resetting it across runs, the second run gets `_x2`, `_x3`,
  // and the evidence-body raw IR differs even though the canonical FOL
  // is the same. The propertyHash (canonical, de-Bruijn-indexed) is
  // unaffected, but the envelope CID hashes the raw evidence body.
  _resetQuantifierCounter();
}

// ---------------------------------------------------------------------------
// describe() / it() — nested-context property declaration sugar.
//
// Mirrors vitest / jest / mocha. Authors organize invariants in nested
// describe blocks; it() declares a single named property whose full
// name is the path through the describe tree.
//
//   describe("parseInt", () => {
//     it("canReturnZero",
//       exists(StringSort, s => eq(parseInt(s), num(0)))
//     );
//
//     describe("non-negative integers", () => {
//       it("round-trip",
//         forAll(Int, n =>
//           implies(gte(n, num(0)), eq(parseInt(toString(n)), n))
//         )
//       );
//     });
//   });
//
// Yields property names: "parseInt > canReturnZero" and
// "parseInt > non-negative integers > round-trip".
// ---------------------------------------------------------------------------

let describePath: string[] = [];

/**
 * Open a named describe block. The body runs immediately; any property()
 * or it() calls inside register with the describe path prepended.
 *
 * Nesting is supported. The path uses " > " as a separator.
 */
export function describe(name: string, body: () => void): void {
  describePath.push(name);
  try {
    body();
  } finally {
    describePath.pop();
  }
}

/**
 * Declare a named invariant. The active describe path is used as a
 * prefix for the full name.
 *
 * The verb `must` is non-negotiable: invariants are obligations, not
 * observations. `it("returns zero")` reads as a test ("it does this");
 * `must("never throw on empty input")` reads as a constraint ("this
 * is required"). The framework writes invariants in the obligation
 * register; the API forces that register at call sites.
 */
export function must(name: string, formula: IrFormula): void {
  const fullName =
    describePath.length === 0 ? name : `${describePath.join(" > ")} > ${name}`;
  property(fullName, formula);
}

/** Skip an invariant. The declaration is acknowledged but not collected. */
must.skip = function (name: string, _formula: IrFormula): void {
  void name;
};

/** Skip a describe block. */
describe.skip = function (name: string, _body: () => void): void {
  void name;
};

// ---------------------------------------------------------------------------
// property() — declares a named property whose body is an IrFormula
// ---------------------------------------------------------------------------

/**
 * Declare a named property with an IR formula body.
 *
 * @param name — the property's identifier (used for diagnostics + memento naming)
 * @param formula — the IR formula constructed via the kit's symbolic primitives
 */
export function property(name: string, formula: IrFormula): void {
  if (activeCollector === null) {
    throw new Error(
      `property("${name}", ...) called outside an active collector. ` +
      `If you're running this file standalone for testing, call beginCollecting() first.`,
    );
  }
  activeCollector.push({ kind: "property", name, formula });
}

// ---------------------------------------------------------------------------
// bridge() — declares that a host-language symbol bridges to a deeper-layer
// published contract by CID.
// ---------------------------------------------------------------------------

export interface BridgeSpec {
  sourceSymbol: string;
  sourceLayer: string;
  targetContractCid: string;
  targetLayer: string;
  notes?: string;
}

/**
 * Declare a bridge from a host-language symbol to a deeper-layer contract.
 *
 * @param name — the bridge's identifier
 * @param spec — the bridge specification (source, target, optional notes)
 */
export function bridge(name: string, spec: BridgeSpec): void {
  if (activeCollector === null) {
    throw new Error(
      `bridge("${name}", ...) called outside an active collector.`,
    );
  }
  activeCollector.push({
    kind: "bridge",
    name,
    sourceSymbol: spec.sourceSymbol,
    sourceLayer: spec.sourceLayer,
    targetContractCid: spec.targetContractCid,
    targetLayer: spec.targetLayer,
    ...(spec.notes !== undefined ? { notes: spec.notes } : {}),
  });
}

// ---------------------------------------------------------------------------
// Quantifier wrappers — exported here so the symbolic module is the
// single import point. (These delegate to the existing IR library's
// builders.)
// ---------------------------------------------------------------------------

import {
  forAll as _forAll,
  exists as _exists,
  _resetCounter as _resetQuantifierCounter,
} from "../quantifiers.js";

export function forAll(sort: Sort, body: (x: IrTerm) => IrFormula): IrFormula {
  return _forAll(sort, body);
}

export function exists(sort: Sort, body: (x: IrTerm) => IrFormula): IrFormula {
  return _exists(sort, body);
}

// ---------------------------------------------------------------------------
// Connectives
// ---------------------------------------------------------------------------

import {
  and as _and,
  or as _or,
  not as _not,
  implies as _implies,
  iff as _iff,
} from "../connectives.js";

export const and = _and;
export const or = _or;
export const not = _not;
export const implies = _implies;
export const iff = _iff;

File: kits/rust/provekit-ir-symbolic/src/property.rs
//! Property and bridge collectors. Mirrors `src/ir/symbolic/property.ts`.
//!
//! Thread-local active collector. `begin_collecting()` opens it; `must()`,
//! `property()`, and `bridge()` push declarations; the returned
//! `FinishHandle::finish()` closes the collector and returns the captured
//! list. The collector is non-reentrant: opening one while another is
//! active is an error.

use std::cell::RefCell;

use serde::{Deserialize, Serialize};

use crate::types::IrFormula;

// ---------------------------------------------------------------------------
// Declaration types
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "kind")]
pub enum Declaration {
    #[serde(rename = "property")]
    Property {
        name: String,
        formula: IrFormula,
    },
    #[serde(rename = "bridge")]
    Bridge {
        name: String,
        #[serde(rename = "sourceSymbol")]
        source_symbol: String,
        #[serde(rename = "sourceLayer")]
        source_layer: String,
        #[serde(rename = "targetContractCid")]
        target_contract_cid: String,
        #[serde(rename = "targetLayer")]
        target_layer: String,
        #[serde(skip_serializing_if = "Option::is_none")]
        notes: Option<String>,
    },
}

impl Declaration {
    pub fn name(&self) -> &str {
        match self {
            Declaration::Property { name, .. } | Declaration::Bridge { name, .. } => name,
        }
    }
}

#[derive(Debug, Clone)]
pub struct BridgeSpec {
    pub source_symbol: String,
    pub source_layer: String,
    pub target_contract_cid: String,
    pub target_layer: String,
    pub notes: Option<String>,
}

// ---------------------------------------------------------------------------
// Collector state — thread-local
// ---------------------------------------------------------------------------

thread_local! {
    static ACTIVE: RefCell<Option<Vec<Declaration>>> = const { RefCell::new(None) };
    static DESCRIBE_PATH: RefCell<Vec<String>> = const { RefCell::new(Vec::new()) };
}

/// Handle returned by `begin_collecting`. Call `.finish()` to retrieve the
/// captured declarations and clear the active collector.
#[must_use = "the collector remains open until finish() is called"]
pub struct FinishHandle {
    _private: (),
}

impl FinishHandle {
    /// Close the collector and return the captured declarations.
    pub fn finish(self) -> Vec<Declaration> {
        ACTIVE.with(|cell| {
            cell.borrow_mut().take().unwrap_or_default()
        })
    }
}

/// Begin collecting declarations. Returns a handle whose `.finish()` ends
/// collection. Non-reentrant: panics if a collection is already active.
pub fn begin_collecting() -> FinishHandle {
    ACTIVE.with(|cell| {
        let mut slot = cell.borrow_mut();
        if slot.is_some() {
            panic!(
                "begin_collecting: another collection is already active; lifting is not re-entrant"
            );
        }
        *slot = Some(Vec::new());
    });
    FinishHandle { _private: () }
}

/// Test helper: clear any active collector and reset describe path +
/// quantifier counter. Use only in test setup/teardown.
pub fn _reset_collector() {
    ACTIVE.with(|cell| { *cell.borrow_mut() = None; });
    DESCRIBE_PATH.with(|cell| { cell.borrow_mut().clear(); });
    crate::quantifiers::_reset_counter();
}

// ---------------------------------------------------------------------------
// describe / must / property / bridge
// ---------------------------------------------------------------------------

/// Open a named describe block. Body runs immediately; nested `must()`
/// calls register with the describe path prepended (joined by " > ").
pub fn describe<F: FnOnce()>(name: &str, body: F) {
    DESCRIBE_PATH.with(|cell| cell.borrow_mut().push(name.to_string()));
    // Use catch_unwind-equivalent? No — keep it simple: panic propagates,
    // and we rely on the caller resetting via _reset_collector if needed.
    // We DO want to pop the path even on panic; use a guard.
    struct Guard;
    impl Drop for Guard {
        fn drop(&mut self) {
            DESCRIBE_PATH.with(|cell| {
                cell.borrow_mut().pop();
            });
        }
    }
    let _g = Guard;
    body();
}

/// describe.skip equivalent — body never runs.
pub fn describe_skip<F: FnOnce()>(_name: &str, _body: F) {
    // intentionally drops the body without invoking it
}

fn current_full_name(name: &str) -> String {
    DESCRIBE_PATH.with(|cell| {
        let path = cell.borrow();
        if path.is_empty() {
            name.to_string()
        } else {
            format!("{} > {}", path.join(" > "), name)
        }
    })
}

/// Declare a named obligation. The active describe path prefixes the name.
pub fn must(name: &str, formula: IrFormula) {
    let full = current_full_name(name);
    property(&full, formula);
}

/// Skip an obligation (no-op).
pub fn must_skip(_name: &str, _formula: IrFormula) {}

/// Declare a property directly (no describe-path prefix applied).
pub fn property(name: &str, formula: IrFormula) {
    ACTIVE.with(|cell| {
        let mut slot = cell.borrow_mut();
        match slot.as_mut() {
            Some(decls) => decls.push(Declaration::Property {
                name: name.to_string(),
                formula,
            }),
            None => panic!(
                "property(\"{}\", ...) called outside an active collector. \
                 Call begin_collecting() first.",
                name
            ),
        }
    });
}

/// Declare a bridge from a host-language symbol to a deeper-layer contract.
pub fn bridge(name: &str, spec: BridgeSpec) {
    ACTIVE.with(|cell| {
        let mut slot = cell.borrow_mut();
        match slot.as_mut() {
            Some(decls) => decls.push(Declaration::Bridge {
                name: name.to_string(),
                source_symbol: spec.source_symbol,
                source_layer: spec.source_layer,
                target_contract_cid: spec.target_contract_cid,
                target_layer: spec.target_layer,
                notes: spec.notes,
            }),
            None => panic!(
                "bridge(\"{}\", ...) called outside an active collector.",
                name
            ),
        }
    });
}

// ---------------------------------------------------------------------------
// Macro sugar
// ---------------------------------------------------------------------------

/// `must!("name", formula)` — sweetened call to `must`.
#[macro_export]
macro_rules! must {
    ($name:expr, $formula:expr) => {
        $crate::property::must($name, $formula)
    };
}

/// `describe!("name", { body })` — sweetened call to `describe`.
/// The body block runs in an `FnOnce` closure.
#[macro_export]
macro_rules! describe {
    ($name:expr, $body:block) => {
        $crate::property::describe($name, || $body)
    };
}

/// `bridge!("name", BridgeSpec { ... })` — sweetened call to `bridge`.
#[macro_export]
macro_rules! bridge {
    ($name:expr, $spec:expr) => {
        $crate::property::bridge($name, $spec)
    };
}
